ftl/common/builder_runner.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A binary for constructing images from a source context."""
+
+import tarfile
+from containerregistry.client import docker_creds
+from containerregistry.client import docker_name
+from containerregistry.client.v2_2 import append
+from containerregistry.client.v2_2 import docker_image
+from containerregistry.client.v2_2 import docker_session
+from containerregistry.client.v2_2 import save
+from containerregistry.transport import transport_pool
+
+import httplib2
+import logging
+
+from ftl.common import cache
+from ftl.common import context
+
+_THREADS = 32
+
+
+class BuilderRunner():
+    def __init__(self, args, builder, cache_version_str):
+        self.args = args
+        self.cache_version_str = cache_version_str
+
+        self._init_transport()
+        self._init_docker_config(args)
+        self._init_ctx(args)
+        self._init_cache(cache_version_str)
+        self.bldr = builder.From(self.ctx)
+
+    def _init_transport(self):
+        self.transport = transport_pool.Http(httplib2.Http, size=_THREADS)
+
+    def _init_docker_config(self, args):
+        # TODO(mattmoor): Support digest base images.
+        self.base_name = docker_name.Tag(args.base)
+        self.base_creds = docker_creds.DefaultKeychain.Resolve(self.base_name)
+
+        self.target_image = docker_name.Tag(args.name)
+        self.target_creds = docker_creds.DefaultKeychain.Resolve(
+            self.target_image)
+
+    def _init_ctx(self, args):
+        self.ctx = context.Workspace(args.directory)
+
+    def _init_cache(self, cache_version_str):
+        self.cash = cache.Registry(
+            self.target_image.as_repository(),
+            self.target_creds,
+            self.transport,
+            cache_version=cache_version_str,
+            threads=_THREADS,
+            mount=[self.base_name])
+
+    def GenerateFTLImage(self):
+        with docker_image.FromRegistry(self.base_name, self.base_creds,
+                                       self.transport) as base_image:
+
+            # Create (or pull from cache) the base image with the
+            # package descriptor installation overlaid.
+            logging.info('Generating dependency layer...')
+            with self.bldr.CreatePackageBase(base_image, self.cash,
```

Comment:
It seems like we could extract the cache getting/setting as well by exposing some methods on builder around the cache key generation. WDYT?